crates/wdl-engine/src/backend/lsf_apptainer/images.rs

```diff
@@ -0,0 +1,234 @@
+//! Management of Apptainer images.
+//!
+//! This module populates and maintains explictly-converted `.sif` files and
+//! reuses them without inducing additional requests to the container registry.
+//!
+//! Apptainer/Singularity has its own image format, `.sif`, but can convert from
+//! Docker-style OCI images more-or-less transparently when images are specified
+//! with a `docker://` prefix. It even caches these conversions so that repeated
+//! invocations of `apptainer exec` with the same image specifier will not
+//! trigger a full fetch and rebuild of the image.
+//!
+//! However, even though the `.sif` images are cached, Apptainer still sends a
+//! request to the container registry when reusing the images in order to ensure
+//! the image is up to date. Depending on the shape of a workflow execution and
+//! the configuration of the container registry, this traffic is enough to cause
+//! sporadic failures in workflow execution, particularly when large numbers of
+//! tasks are invoked with the same WDL `container` requirement.
+//!
+//! By explicitly converting to `.sif` and using those files in `apptainer exec`
+//! invocations instead of `docker://` specifiers, we avoid this additional
+//! container registry traffic. To sidestep the issue of staleness between the
+//! locally-converted `.sif` images and the contents of the container
+//! registry, the `.sif` files are only used for a single workflow execution. We
+//! still benefit from the Apptainer cache when building new `.sif` files, so
+//! this is not much of a slowdown, but it does increase disk space consumption
+//! depending on where the images directory is created.
```

Comment:
I actually like this behavior. I wouldn't want a tagged image to get updated within a workflow, which otherwise could happen if a push happened between task executions.

Reply:
Oh yes, that part is absolutely intended and desirable behavior. What's not ideal is that we don't have a way for the caches to be reused between top-level invocations, and so the current behavior ends up leaving behind a bunch of detritus on disk.

Reply:
I've mentioned this before in passing I believe, but maybe not recently. I can imagine us having a `sprocket clean` command that purges any locations we might cache things. Not needed now of course, but something we may want in the future